server/embed/etcd.go

```diff
@@ -277,6 +277,9 @@ func StartEtcd(inCfg *Config) (e *Etcd, err error) {
 			return e, err
 		}
 	}
+
+	e.Server.SyncLearnerPromotionIfNeeded()
```

Comment:
Not 100% sure about consequences of calling sync here, by proposing to run it during bootstrap I meant within `bootstrap.go`

Reply:
How do you think this change will interact with corrupt check? Even if we do the sync before corrupt check on one member, next member will start with corrupt check before sync.

Reply:
> I don't know if we really don't run any goroutines in background at this point, and proving so might be very hard.

We call `SyncLearnerPromotionIfNeeded` before starting the raft loop, also before starting to serve client & peer requests. It's safe. What's your concern here?



> How do you think this change will interact with corrupt check?

If there is data corrupt, we should leave it as it's.

Reply:
I am thinking if we still need this migration.

I found there is publish API call during starting. Since we already allow to override member information, this publish can help us migration (if I understand it correctly).

https://github.com/etcd-io/etcd/blob/3c65dfa9d3f85fbfff19f5f0e0f57a67ecefa48d/server/etcdserver/server.go#L2132

* If there is snapshot, the v2store contains valid data. UpdateAtribute will override existing invalid one.
* If there is not snapshot, WAL replay can fix this issue as well.

Reply:
@serathius please let me know if you still have any concern on my previous comment. I think we should be all good now.



> I am thinking if we still need this migration.
> 
> I found there is publish API call during starting. Since we already allow to override member information, this publish can help us migration (if I understand it correctly).

@fuweid I don't think we are on the same page. The publish logic is about attribute, what we are resolving/talking about in this PR (and issue 19557) is about IsLearner, which is part of RaftAttribute.

Reply:
> override member information

Yes, this is true. thx for pointing out this. 

Although the publish logic is only supposed to update attribute, but the implementation just overwrites the whole member info into v3store (bbolt), so it can automatically fix the issues which are affected by https://github.com/etcd-io/etcd/issues/19557 on top of the patch https://github.com/etcd-io/etcd/pull/19563 

Just raised another PR with only e2e test cases https://github.com/etcd-io/etcd/pull/19629, which I believe is better than https://github.com/etcd-io/etcd/pull/19627

Reply:
For release-3.6, we don't need any further validation (i.e. requiring 3.5.20+ before upgrading to 3.6) any more, reasons:
- If there are 2 or more (already promoted) learners, then etcd will crash anyway.
- If there is only 1 (already promoted) learner, then etcd will automatically fix it due to the same reason as mentioned above. 

So for release-3.6, we only need to add two e2e test cases:
- the kubeadm style upgrade case
- verify etcd is able to automatically fix the issue when upgrading to 3.6 for the one learner case, similar to https://github.com/etcd-io/etcd/pull/19629